astropy/io/fits/tests/test_table.py

```diff
@@ -3313,6 +3313,55 @@ def test_multidim_VLA_tables(self):
                 hdus[1].data["test"][1], np.array([[0.0, 1.0, 2.0], [3.0, 4.0, 5.0]])
             )
 
+    def test_heterogeneous_VLA_tables(self):
+        """
+        Check the behaviour of heterogeneous VLA object.
+        """
+
+        # The column format fix the type of the arrays in the VLF object.
+        a = np.array([45, 30])
+        b = np.array([11.0, 12.0, 13])
+        var = np.array([a, b], dtype=object)
+
+        c1 = fits.Column(name="var", format="PJ()", array=var)
+        hdu = fits.BinTableHDU.from_columns([c1])
+        assert hdu.data[0].array.dtype[0].subdtype[0] == "int32"
+
+        # Strings in the VLF object can't be added to the table
+        a = np.array([45, "thirty"])
+        b = np.array([11.0, 12.0, 13])
+        var = np.array([a, b], dtype=object)
+
+        c1 = fits.Column(name="var", format="PJ()", array=var)
+        with pytest.raises(
+            ValueError, match=r"invalid literal for int\(\) with base 10"
+        ):
+            fits.BinTableHDU.from_columns([c1])
+
+    def test_write_VLA_tables_with_unified(self):
+        """
+        Write VLA objects with the unified I/O interface.
+        See https://github.com/astropy/astropy/issues/11323
+        """
+
+        # Make a FITS table with a variable-length array column
+        a = np.array([45, 30])
+        b = np.array([11, 12, 13])
+        c = np.array([45, 55, 65, 75])
```

Comment:
Need also to test that an exception is raised with heterogeneous types + maybe test different types etc.